pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,393 @@
+	}
+
+	for _, plugin := range criAuthPluginConfig.Plugins {
+		provider := &pluginProvider{
+			matchImages:          plugin.MatchImages,
+			cache:                cache.NewExpirationStore(cacheKeyFunc, &cacheExpirationPolicy{}),
+			defaultCacheDuration: plugin.DefaultCacheDuration.Duration,
+			plugin: &execPlugin{
+				name:         plugin.Name,
+				pluginBinDir: pluginBinDir,
+				args:         convertArgs(plugin.ExtraArgs),
+			},
+		}
+
+		credentialprovider.RegisterCredentialProvider(plugin.Name, provider)
+	}
+
+	return nil
+}
+
+// pluginProvider is the plugin-based implementation of the DockerConfigProvider interface.
+type pluginProvider struct {
+	sync.Mutex
+
+	// matchImages defines the matching image URLs this plugin should operate against.
+	// The plugin provider will not return any credentials for images that do not match
+	// against this list of match URLs.
+	matchImages []string
+
+	// cache stores DockerConfig entries with an expiration time based on the cache duration
+	// returned from the CRI auth plugin.
+	cache cache.Store
+	// defaultCacheDuration is the default duration credentials are cached in-memory if the auth plugin
+	// response did not provide a cache duration for credentials.
+	defaultCacheDuration time.Duration
+
+	// plugin is the exec implementation of the credential providing plugin.
+	plugin Plugin
+}
+
+// cacheEntry is the cache object that will be stored in cache.Store.
+type cacheEntry struct {
+	key         string
+	credentials credentialprovider.DockerConfig
+	expiresAt   time.Time
+}
+
+// cacheKeyFunc extracts AuthEntry.MatchKey as the cache key function for the plugin provider.
+func cacheKeyFunc(obj interface{}) (string, error) {
+	key := obj.(*cacheEntry).key
+	return key, nil
+}
+
+// cacheExpirationPolicy defines implements cache.ExpirationPolicy, determining expiration based on the expiresAt timestamp.
+type cacheExpirationPolicy struct{}
+
+// IsExpired returns true if the current time is after cacheEntry.expiresAt, which is determined by the
+// cache duration returned from the CRI auth plugin response.
+func (c *cacheExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
+	return time.Now().After(entry.Obj.(*cacheEntry).expiresAt)
+}
+
+// Provide returns a credentialprovider.DockerConfig based on the credentials returned
+// from cache or the exec plugin.
+func (p *pluginProvider) Provide(image string) credentialprovider.DockerConfig {
+	if !p.isImageAllowed(image) {
+		return credentialprovider.DockerConfig{}
+	}
+
+	p.Lock()
+	defer p.Unlock()
+
+	cachedConfig, found, err := p.getCachedCredentials(image)
+	if err != nil {
+		klog.Errorf("Failed to get cached docker config: %v", err)
+		return credentialprovider.DockerConfig{}
+	}
+
+	if found {
+		return cachedConfig
+	}
+
+	// fetch credentials from the exec plugin if the image is allowed and not already matched against cached objects
+	response, err := p.plugin.ExecPlugin(image)
```

Comment:
calls to exec could be expensive... does `Provide` get called in parallel for multiple images? if so, doing this in the critical section under lock could really slow things down

Reply:
If we don't call this under lock, multiple plugins could call `ExecPlugin` in parallel when we only need the first call to `ExecPlugin` and subsequent calls could fetch from cache. I assumed making other calls wait for cache is better than every call execing the plugin.

Reply:
this is fine for alpha... add a note to the list for beta to see if this gets called in parallel, and if so, see if this needs to be addressed

Reply:
ack